ethcore/engines/authority-round/src/randomness.rs

```diff
@@ -0,0 +1,248 @@
+//! method.
+//!
+//! No additional state is kept inside the `RandomnessPhase`, it must be passed in each time.
+//!
+//! The process of generating random numbers is a simple finite state machine:
+//!
+//! ```text
+//!                                                       +
+//!                                                       |
+//!                                                       |
+//!                                                       |
+//! +--------------+                              +-------v-------+
+//! |              |                              |               |
+//! | BeforeCommit <------------------------------+    Waiting    |
+//! |              |          enter commit phase  |               |
+//! +------+-------+                              +-------^-------+
+//!        |                                              |
+//!        |  call                                        |
+//!        |  `commitHash()`                              |  call
+//!        |                                              |  `revealNumber`
+//!        |                                              |
+//! +------v-------+                              +-------+-------+
+//! |              |                              |               |
+//! |  Committed   +------------------------------>    Reveal     |
+//! |              |  enter reveal phase          |               |
+//! +--------------+                              +---------------+
+//! ```
+//!
+//! Phase transitions are performed by the smart contract and simply queried by the engine.
+//!
+//! A typical case of using `RandomnessPhase` is:
+//!
+//! 1. `RandomnessPhase::load()` the phase from the blockchain data.
+//! 2. Call `RandomnessPhase::advance()`.
+//!
+//! A production implementation of a randomness contract can be found here:
+//! https://github.com/poanetwork/posdao-contracts/blob/4fddb108993d4962951717b49222327f3d94275b/contracts/RandomAuRa.sol
+
+use derive_more::Display;
+use ethabi::Hash;
+use ethabi_contract::use_contract;
+use ethereum_types::{Address, H256, U256};
+use keccak_hash::keccak;
+use log::{debug, error};
+use parity_crypto::publickey::{ecies, Error as CryptoError};
+use parity_bytes::Bytes;
+use rand::Rng;
+use engine::signer::EngineSigner;
+
+use crate::util::{BoundContract, CallError};
+
+/// Random number type expected by the contract: This is generated locally, kept secret during the commit phase, and
+/// published in the reveal phase.
+pub type RandNumber = H256;
+
+use_contract!(aura_random, "../../res/contracts/authority_round_random.json");
+
+/// Validated randomness phase state.
+#[derive(Debug)]
+pub enum RandomnessPhase {
+	// NOTE: Some states include information already gathered during `load` (e.g. `our_address`,
+	//       `round`) for efficiency reasons.
+	/// Waiting for the next phase.
+	///
+	/// This state indicates either the successful revelation in this round or having missed the
+	/// window to make a commitment.
```

Comment:
Maybe dumb q: what is the window? A range of blocks? Where do we set this? In the `aura_random` contract? (If yes, I think it's a good idea to add notes for any piece of data that is defined in the contract as opposed to the rust code).

Reply:
That's up to the contract. In POSDAO it's a range of blocks, yes.
I'll add a comment.

Reply:
The comment was added in https://github.com/poanetwork/parity-ethereum/commit/4799aeb3d11bac2910ff04e5714631a541d92aec.